### PR TITLE
Relocate GroupInfo into shared process-group utilities

### PR DESCRIPTION
**Pull request by fzyzcjy**
Move the GroupInfo dataclass out of training_utils/parallel.py into a shared
process_group_utils module that also provides multi-process-group helpers
(GeneralPGUtil / MultiPGUtil) for collectives over single or hierarchical
process groups, used by the cross-replica / effective-DP code paths.

- process_group_utils.py: GroupInfo + GeneralPGUtil / MultiPGUtil (+ tests).
- training_utils/parallel.py: import GroupInfo from the shared module.
- megatron / fsdp parallel.py: import GroupInfo from the shared module.
- distributed_utils.py: use GeneralPGUtil for masked-whiten all-reduce.

### CODE REVIEW

File: miles/utils/process_group_utils.py [hunk 1]
```diff
@@ -0,0 +1,345 @@
...
+        # Must .wait(): torchft arms its per-collective timeout only inside Work.wait();
+        # a fire-and-forget barrier hangs a dead-peer cell until the NCCL watchdog kills it.
+        _check_wait(group.barrier(), "barrier")
+
+    def all_gather(
+        self, output_tensors: list[torch.Tensor], input_tensor: torch.Tensor, group: dist.ProcessGroup
+    ) -> None:
+        # AllgatherOptions is not re-exported by torch.distributed (unlike
+        # AllreduceOptions, BroadcastOptions, GatherOptions). PyTorch omission.
+        from torch._C._distributed_c10d import AllgatherOptions
```
Comment (Shi-Dong) [inline miles/utils/process_group_utils.py]: Please hoist this import to the beginning

File: miles/utils/process_group_utils.py [hunk 2]
```diff
@@ -0,0 +1,345 @@
...
+    """
+    # --- Begin: adapted from PyTorch v2.11.0 gather_object ---
+
+    my_group_rank = util.get_rank(group)  # was: group.rank()
+    if my_group_rank == 0:
+        assert object_gather_list is not None
+    else:
+        assert object_gather_list is None
+
+    from torch.distributed.distributed_c10d import _get_object_coll_device
```
Comment (Shi-Dong) [inline miles/utils/process_group_utils.py]: Please hoist this import to the beginning

File: miles/backends/experimental/fsdp_utils/parallel.py [hunk 3]
```diff
@@ -6,7 +6,9 @@
 
 from miles.utils.distributed_utils import get_gloo_group
 
-from ...training_utils.parallel import GroupInfo, ParallelState
+from miles.utils.process_group_utils import GroupInfo
+
+from ...training_utils.parallel import ParallelState
```
Comment (Shi-Dong) [inline miles/backends/experimental/fsdp_utils/parallel.py:11]: Please switch relative import to absolute import.

File: miles/backends/megatron_utils/parallel.py [hunk 4]
```diff
@@ -8,7 +8,9 @@
 from megatron.core.utils import get_model_config
 from megatron.training.global_vars import get_args
 
-from ..training_utils.parallel import GroupInfo, ParallelState, get_parallel_state
+from miles.utils.process_group_utils import GroupInfo
+
+from ..training_utils.parallel import ParallelState, get_parallel_state
```
Comment (Shi-Dong) [inline miles/backends/megatron_utils/parallel.py:13]: Please switch relative import to absolute import.

File: miles/utils/process_group_utils.py [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+    def gather(
+        self,
+        input_tensor: torch.Tensor,
+        gather_list: list[torch.Tensor] | None,
+        group: dist.ProcessGroup,
+    ) -> None:
+        dist.gather(input_tensor, gather_list=gather_list, dst=dist.get_global_rank(group, 0), group=group)
+
+    def gather_object(self, obj: Any, object_gather_list: list[Any] | None, group: dist.ProcessGroup) -> None:
+        dist.gather_object(obj, object_gather_list, dst=dist.get_global_rank(group, 0), group=group)
```
[the remark this answers is not in the document]
Comment (Shi-Dong) [reply miles/utils/process_group_utils.py:149]: Please apply.

File: miles/utils/process_group_utils.py [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+        object_gather_list[i] = _tensor_to_object(tensor, tensor_size, group)
+
+    # --- End: adapted from PyTorch v2.11.0 gather_object ---
+
+
+def collective_bool_and(*, value: bool, group: dist.ProcessGroup) -> bool:
+    """Make a bool `and` operation on all ranks in this process group"""
+    tensor = torch.tensor([1.0 if value else 0.0], dtype=torch.float32)
+    GeneralPGUtil.create(group).all_reduce(tensor, group, op=dist.ReduceOp.MIN)
+    return tensor.item() > 0.5
```
[the remark this answers is not in the document]
Comment (Shi-Dong) [reply miles/utils/process_group_utils.py:343]: Please apply.

**Review by Shi-Dong — COMMENTED** [review]
Posted a couple of nits, but please also apply Gemini's suggestions as they look valid.

**Review by yueming-yuan — APPROVED** [review]
